src/sort.py

```diff
@@ -0,0 +1,148 @@
+                i -= 1
+            else:
+                candidate_paths = {s2}
+                if i > n and s1[i]:
+                    candidate_paths |= s1[i].candidate_paths
+                if s1[i-1]:
+                    candidate_paths |= s1[i-1].candidate_paths
+                index.append(Profile(s2.nodes[j-1], [s2], candidate_paths, s2.nodes[j-1].node.id in prev))
+                prev.add(s2.nodes[j-1].node.id)
+                j -= 1
+
+        while i > 0:
+            prev_paths = s1[i - 1].paths
+            prev_candidates = s1[i-1].candidate_paths
+            index.append(Profile(s1[i - 1].node, prev_paths, prev_candidates, s1[i - 1].node.node.id in prev))
+            prev.add(s1[i - 1].node.node.id)
+            i -= 1
+
+        while j > 0:
+            prev.add(s2.nodes[j - 1].node.id)
+            index.append(Profile(s2.nodes[j - 1], [s2], {s2}, False))
+            j -= 1
+
+        index.reverse()
+
+        return index
+
+    def to_slices(self, profile: List[Profile]) -> List[Path]:
+        factory_input = []
+        current_slice = Slice([])
+        current_paths = []
+
+        for index, prof in enumerate(profile):
+            paths = [x for x in prof.paths]
+            all_path_set = set([x for x in current_paths])
+            # print(prof, current_slice, current_paths)
+            candidate_paths_set = prof.candidate_paths
+            if index + 1 != len(profile):
+                candidate_paths_set |= profile[index+1].candidate_paths
+
+            if len(prof.paths) == len(candidate_paths_set):
+                if len(current_slice.nodes) > 0:
+                    if prof.candidate_paths - all_path_set != set():
+                        current_slice.add_node(Node("", prof.candidate_paths - all_path_set))
+                    factory_input.append(current_slice)
+                factory_input.append(Slice([Node(prof.node.node.seq, paths, prof.node.node.id)]))
+                current_slice = Slice([])
+                current_paths = []
+            else:
+                if set([x for x in prof.paths]) & all_path_set != set():
+                    if len(current_slice.nodes) > 0:
+                        if prof.candidate_paths - all_path_set != set():
+                            current_slice.add_node(Node("", prof.candidate_paths - all_path_set))
+                        factory_input.append(current_slice)
+                    current_slice = Slice([Node(prof.node.node.seq, paths, prof.node.node.id)])
+                    current_paths = paths
+                else:
+                    current_slice.add_node(Node(prof.node.node.seq, paths, prof.node.node.id))
+                    current_paths.extend(paths)
+
+        if len(current_slice.nodes) > 0:
+            all_path_set = set([x for x in current_paths])
+            if profile[-1].candidate_paths - all_path_set != set():
+                current_slice.add_node(Node("", prof.candidate_paths - all_path_set))
```

Comment:
Pycharm informs me: By the indentation, prof should no longer be in scope.

Reply:
hmmmm, I will check the codes.